server/api/crud/model_monitoring/deployment.py

```diff
@@ -516,84 +679,150 @@ def _check_if_already_deployed(
 
-        func.set_label(
-            mm_constants.ModelMonitoringAppLabel.KEY,
-            mm_constants.ModelMonitoringAppLabel.VAL,
+    def _reuse_image(
+        self,
+        function: mlrun.runtimes.RemoteRuntime,
+        name: str,
+        base_image: str,
+    ):
+        # Get existing image of the controller function
+        nuclio_image = self._get_existing_nuclio_image(
+            name=name,
+            base_image=base_image,
         )
+        if nuclio_image:
+            self._set_nuclio_image_config(function, nuclio_image)
 
-        fn, ready = server.api.utils.functions.build_function(
-            db_session=self.db_session, auth_info=self.auth_info, function=func
-        )
+    def _get_existing_nuclio_image(self, name: str, base_image: str):
+        """Get the nuclio image that was built with the same base image for the given function name"""
+        version_hash_key = self._get_version_hash_key()
 
-        logger.debug(
-            "Submitted the histogram data drift app deployment",
-            app_data=fn.to_dict(),
-            app_ready=ready,
-        )
+        # if the image manifest is empty, read the image manifest as it might be updated
+        if version_hash_key not in self._image_manifest:
+            self._read_image_manifest()
 
-    def is_monitoring_stream_has_the_new_stream_trigger(self) -> bool:
-        """
-        Check if the monitoring stream function has the new stream trigger.
+        function_image_info = self._image_manifest.get(version_hash_key, {}).get(name)
 
-        :return: True if the monitoring stream function has the new stream trigger, otherwise False.
-        """
+        # we want to get the nuclio image that was built with the same base image
+        if function_image_info and base_image in function_image_info:
+            return function_image_info.get(base_image)
 
-        try:
-            function = server.api.crud.Functions().get_function(
-                name=mm_constants.MonitoringFunctionNames.STREAM,
-                db_session=self.db_session,
-                project=self.project,
-            )
-        except mlrun.errors.MLRunNotFoundError:
-            logger.info(
-                "The stream function is not deployed yet when the user will run `enable_model_monitoring` "
-                "the stream function will be deployed with the new & the old stream triggers",
-                project=self.project,
-            )
-            return True
+        return None
 
-        if (
-            function["spec"]["config"].get(
-                f"spec.triggers.monitoring_{mm_constants.MonitoringFunctionNames.STREAM}_trigger_1"
-            )
-            is None
-        ):
-            logger.info(
-                "The stream function needs to be updated with the new stream trigger",
-                project=self.project,
-            )
-            return False
-        return True
+    def _update_image_manifest(
```

Comment:
I'd like this method to work parallel. Writing to the manifest file parallel is a bad practice. But you can wait for the nuclio deployments in parallel. Wait for all the statuses to arrive and then write to the file. (Unless we make these rows in a DB table in which case everything can be parallel)